[Docs] Terrain Developer Guide (#2122)
o3de/o3de.org: content/docs/user-guide/visualization/environments/terrain/terrain-developer-guide/future-work.md

---
linktitle: Future Work
title: Future Work
description: Possible future improvements to the terrain system.
weight: 500
toc: true
---

> There are a number of improvements that could potentially be made to the terrain system for anyone interested in taking the system further.

Author: ```suggestion
There are a number of improvements that can potentially be made to the terrain system for anyone interested in taking the system further. You can reach out to the O3DE community to talk through any terrain enhancements or ideas on the #sig-content channel on [Discord ](https://discord.com/invite/o3de). We are happy to help!
```

Author: How's that sound?

Reviewer: Hm, I was hoping my editing on your suggestion would show so you can see it. (I noticed that works in GH issues.) So apologies for lack of transparency there. I made two small changes, changed "talk through" to "talk about", and removed a space after "Discord".

## Large-scale improvements

This category of improvements requires architecting new solutions to provide extensive new capabilities.

### 3D geometry support

> Instead of a heightfield-based approach, the APIs and renderer could be extended to use a full 3D approach, whether it's through voxels, displacement maps, mesh ingestion, raytracing, or some other technology. This would enable the system to support caves, arches, cliff geometry, and other vertical or overlapping terrain features that simply can't be supported through a pure heightfield approach.

Author: For my edification, can you explain the use of "will" instead of "would" here? Since this is a hypothetical scenario that hasn't happened yet, isn't "would" the correct choice?

i.e. "If I extend it, it would support caves" vs "If I extend it, it will support caves".

> Instead of a heightfield-based approach, the APIs and renderer could be extended to use a full 3D approach, whether it's through voxels, displacement maps, mesh ingestion, raytracing, or some other technology. This would enable the system to support caves, arches, cliff geometry, and other vertical or overlapping terrain features that simply can't be supported through a pure heightfield approach.

Reviewer: Oh yes, sorry for the miss. Given the hypothetical or conditional statement, "would" is more accurate.

> The API extensions would likely consist of redefining the current APIs to provide the _highest elevation_ terrain data that appears below the given query input positions, and then adding more APIs to provide _all_ the terrain data that exists at or below the given query positions.

Author: Are you sure? I think "may" means "possible", but doesn't imply "likely". (The definition of "may" is _literally_ just "expressing possibility.")

If the goal is to remove "would", how about "will likely" instead of "would likely"? This is one place that I can understand the would/will swap, since the addition of "likely" keeps it in the hypothetical instead of the absolute.

> The API extensions would likely consist of redefining the current APIs to provide the _highest elevation_ terrain data that appears below the given query input positions, and then adding more APIs to provide _all_ the terrain data that exists at or below the given query positions.

Reviewer: Yeah, at least compared to "might", "may" indicates greater chance of possibility. I understand that probably isn't intuitive for readers though. "will likely" is perfectly fine

### GPU calculation support

> The terrain system is built on top of the gradient components, which perform all of their calculations CPU-side. However, the types of calculations being performed are ideal for GPU-side calculation. An overhaul or replacement of the gradient system that supports GPU-side calculations could provide orders of magnitude improvements to the terrain system. The primary caveats are that the data is still needed CPU-side to feed to physics, generalized raycasts, and other systems, and that it should be possible to still use the terrain system on devices with limited or no GPU capabilities (low-end phones, headless servers, etc).

Reviewer: For various reasons, I suggest using a simpler adjective, rather than orders of magnitude. The primary reason is that it's long and I think breaks the reading flow. 
```suggestion
The terrain system is built on top of the gradient components, which perform all of their calculations on the CPU-side. However, these calculations can be computed more optimally on the GPU-side. An overhaul or replacement of the gradient system that supports GPU-side calculations can provide significant improvements to the terrain system's performance. The primary caveats are that the data is still needed on the CPU-side to feed to physics, generalized raycasts, and other systems, and that it should be possible to still use the terrain system on devices with limited or no GPU capabilities (low-end phones, headless servers, etc).
```

Author: How about:
```suggestion
The terrain system is built on top of the gradient components, which perform all of their calculations on the CPU. However, a GPU can perform bulk computations like these far more optimally. An overhaul or replacement of the gradient system that supports GPU-side calculations can improve the terrain system's performance by orders of magnitude. The primary caveats are that the data is still needed on the CPU to feed to physics, generalized raycasts, and other systems, and that it should be possible to still use the terrain system on devices with limited or no GPU capabilities (low-end phones, headless servers, etc).
```

The phrase "orders of magnitude" is more specific and objective than "significant" so I'd like to preserve it.

Reviewer: I wasn't sure if it was figurative or not, so in that case, agree to keep orders of magnitude, and your edit looks good.

## Additional features

> This category of improvements should be possible to incrementally add into the system without fundamentally altering anything about how the system currently works.

Reviewer: ```suggestion
This category of improvements may be incrementally added to the system without fundamentally altering anything about how the system currently works.
```

### Triangle split direction

> The terrain currently always triangulates each quad in the same uniform direction. This could be replaced with heuristics for choosing the best direction for each quad. This could even be exposed as multiple heuristics - same direction for every quad, alternating directions per quad, best choice per quad based on vertex heights. This decision should be controlled at the terrain system level, so that the information can be provided consistently to the terrain debugger, terrain physics, terrain rendering, terrain raycasting, etc.

Reviewer: ```suggestion
Currently, the terrain always triangulates each quad in the same uniform direction. This can be replaced with heuristics for choosing the best direction for each quad. This can even be exposed as multiple heuristics: same direction for every quad, alternating directions per quad, or best choice per quad based on vertex heights. This decision should be controlled at the terrain system level, so that the information can be provided consistently to the terrain debugger, terrain physics, terrain rendering, terrain raycasting, etc.
```

### Improved shape to height workflows

> It's currently somewhat non-intuitive to create terrain heights directly from primitive shape components. The best options are to use either the Shape Falloff Gradient or the Surface Altitude Gradient, but they're both currently problematic. The Shape Falloff Gradient creates falloff based on distance from the bottom of a box, not just simply the position of the shape within a box, so it's hard to control. The Surface Altitude Gradient doesn't support falloff and the auto-refresh doesn't work. Either of these components could be improved, or a new component could be added, to make it easier to just place a shape with falloff into the world and turn it into a height gradient.

Reviewer: ```suggestion
Currently, it's somewhat non-intuitive to create terrain heights directly from primitive shape components. The best options are to use either the Shape Falloff Gradient or the Surface Altitude Gradient, but they're both problematic at this moment. The Shape Falloff Gradient creates falloff based on distance from the bottom of a box, not simply the position of the shape within a box, making it hard to control. The Surface Altitude Gradient doesn't support falloff and the auto-refresh doesn't work. Either of these components can be improved or a new component can be added, to make it easier to place a shape with falloff into the world and turn it into a height gradient.
```

### Masking / blending

> Add controls to the Terrain Layer Spawner and the Terrain Macro Material to allow each one to mask and blend instead of completely replacing the data. This would make it much easier to create small terrain "stamps" for things like craters.

Reviewer: ```suggestion
Add controls to the Terrain Layer Spawner and the Terrain Macro Material to allow each one to mask and blend instead of completely replacing the data. This will make it easier to create small terrain "stamps" for things like craters.
```

### First-class "hole" support

> Holes can be created right now by authoring a high-priority terrain layer spawner with no ground plane for whatever size is desired. However, it would be nice to have additional authoring controls to make it possible to put holes directly into gradient data, possibly either through a second alpha channel, or through a separate gradient on the Terrain Height Gradient List component.

Reviewer: ```suggestion
Holes can be created right now by authoring a high-priority terrain layer spawner with no ground plane for whatever size is desired. However, it will be nice to have additional authoring controls to make it possible to put holes directly into gradient data, possibly either through a second alpha channel, or through a separate gradient on the Terrain Height Gradient List component.
```

### Improved terrain texturing

> There are a number of features that would be nice to add to the terrain texturing:

Reviewer: ```suggestion
There are a number of features that will be nice to add to the terrain texturing:
```
